File: backend/src/models/Vocabulary.js
const mongoose = require('mongoose');

const vocabularySchema = new mongoose.Schema(
  {
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true, index: true },
    word: { type: String, required: true, trim: true },
    meaning: { type: String, trim: true },
    sourceLanguage: { type: String, trim: true },
    targetLanguage: { type: String, trim: true },
    // Nguồn thêm từ: từ đoạn hội thoại (click chọn) hoặc tự nhập tay
    source: { type: String, enum: ['conversation', 'manual'], default: 'manual' },
    // Đánh dấu đã thêm vào Flashcard để luyện tập chưa
    inFlashcard: { type: Boolean, default: false },
    audioUrl: { type: String }, // âm thanh phát âm lấy từ từ điển online
  },
  { timestamps: true }
);

vocabularySchema.index({ user: 1, word: 1, targetLanguage: 1 }, { unique: true });

module.exports = mongoose.model('Vocabulary', vocabularySchema);

File: mobile/lib/models/vocabulary.dart
class VocabularyItem {
  final String id;
  final String word;
  final String? meaning;
  final String? sourceLanguage;
  final String? targetLanguage;
  final String source; // 'conversation' | 'manual'
  final bool inFlashcard;
  final String? audioUrl;
  final String? phonetic; // phiên âm (dấu phát âm), vd: /wɜːrd/
  // Chỉ có ý nghĩa với từ trong Kho hệ thống: từ này đã có trong sổ tay của tôi chưa
  final bool inMyVocabulary;

  VocabularyItem({
    required this.id,
    required this.word,
    this.meaning,
    this.sourceLanguage,
    this.targetLanguage,
    this.source = 'manual',
    this.inFlashcard = false,
    this.audioUrl,
    this.phonetic,
    this.inMyVocabulary = false,
  });

  factory VocabularyItem.fromJson(Map<String, dynamic> json) {
    return VocabularyItem(
      id: json['_id'] ?? '',
      word: json['word'] ?? '',
      meaning: json['meaning'],
      sourceLanguage: json['sourceLanguage'],
      targetLanguage: json['targetLanguage'],
      source: json['source'] ?? 'manual',
      inFlashcard: json['inFlashcard'] ?? false,
      audioUrl: json['audioUrl'],
      phonetic: json['phonetic'],
      inMyVocabulary: json['inMyVocabulary'] ?? false,
    );
  }
}
